### DISCUSSION TITLE
Allow callable as argument to "lazy" for orm.relationship

### DISCUSSION THREAD

**Discussion by the-wondersmith** [sqlalchemy/sqlalchemy#7277] [Ideas]
I'm currently working on an application whose core functionality is driven by SQLAlchemy and makes heavy use of `sqlalchemy.orm.relationship`. 

The type annotation for the `lazy` argument looks something like:

```python
from typing import Literal, Optional, Union

LOAD_STRATEGY = Union[
    Optional[bool],
    Literal[
        "select",
        "immediate",
        "joined",
        "subquery",
        "selectin",
        "noload",
        "raise",
        "raise_on_sql",
        "dynamic",
    ],
]
```

It would be very helpful if instead it could be more like:

```python
from typing import Callable, Literal, Optional, Union
from sqlalchemy.orm import RelationshipProperty

STRATEGY_VALUE = ...  # see above
LOAD_STRATEGY = Union[Callable[[RelationshipProperty], STRATEGY_VALUE], STRATEGY_VALUE]
```

Quite a few things support late evaluation in SQLAlchemy, but the load strategy for relationships does not (currently). To make a very long winded story much shorter, it would be extremely useful for things like profiling/optimization and troubleshooting if the `lazy` keyword argument also accepted a late-evaluated callable that was expected to return the proper value for the desired load strategy.

I've written and tested the code for the relevant change (it's literally just one line). I did [open a PR ](https://github.com/sqlalchemy/sqlalchemy/pull/7273) for it, but Mike closed it because I (admittedly) didn't follow proper procedure for doing so. I'm trying (and hopefully succeeding) to do so now.

What are the next steps / requirements to proceed?

**Comment by CaselIT** [comment]
Hi,

What is the use case you are trying to solve?
`lazy` is a static element that does not changes once the model has been changes, so I guess it's an unusual request. Could you share move?

**Reply by CaselIT** [reply]
btw this probably fitted an issue - use case better compared to a discussion, but it's fine here also

**Reply by the-wondersmith** [reply]
Absolutely! In this specific case, the issue we're trying to solve revolves around profiling and trying to optimize load on the DB server by tweaking the way models pull their related records when they're queried. Due to the way the application is built/deployed it's _way_ easier to update the environment variables than it is to update the application itself, so having the relationship load strategies be configurable via environment variables (or other dynamic means) is extremely helpful in rapidly checking the impact of a given configuration.

I know that `lazy` is a static element that does not changes once the mappers have been configured. Even so, lots of other things in SQLAlchemy allow a callable to be supplied as an argument and then evaluate that callable later on. [The change](https://github.com/sqlalchemy/sqlalchemy/pull/7273/files#diff-2b15a73b73756a1c2e3aea07ed29efe2b3978b98367190bb850ff437720e4d45R700) to allow `lazy` to do so as well is minimal to say the least, so I didn't figure anyone would really object to it even if the perceived "value" wasn't all that high.

**Reply by CaselIT** [reply]
Usually callables and other means of configuration are provided because it may refer to another table or model, that may be impossible to pass directly because of circular imports issues or late evaluation.

From your use case, can't you just call the function when defining the model? Environment variable are static (from the outside of the process) once the application has started so it should not matter if it's loaded at module import time or at mapped configuration.

Btw the problem by itself is not really the amount of code that's added, it's that sqlalchemy and relationship in particular, already has a large number of options and ways of configuring itself (many would probably say it has too many :) ) so we should try to add new options only if there is a clear use case for them

@zzzeek thoughts on this one?
